// @chakra-ui/core/dist/useNumberInput/utils.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.isNumberKey = isNumberKey;
exports.preventNonNumberKey = preventNonNumberKey;
exports.roundToPrecision = roundToPrecision;
exports.calculatePrecision = calculatePrecision;

var _prettyNum = _interopRequireWildcard(require("pretty-num"));

function isNumberKey(event) {
  var charCode = event.which ? event.which : event.keyCode;
  if (event.key === ".") return true;
  if (charCode > 31 && (charCode < 48 || charCode > 57)) return false;
  return true;
}

function preventNonNumberKey(event) {
  if (!isNumberKey(event)) {
    event.preventDefault();
  }
}

function roundToPrecision(value, precision) {
  return (0, _prettyNum["default"])(value, {
    precision: precision,
    precisionSetting: _prettyNum.PRECISION_SETTING.FIXED
  });
}

function calculatePrecision(value) {
  var groups = /[1-9]([0]+$)|\.([0-9]*)/.exec(String(value));

  if (!groups) {
    return 0;
  }

  if (groups[1]) {
    return -groups[1].length;
  }

  if (groups[2]) {
    return groups[2].length;
  }

  return 0;
}
